### DISCUSSION TITLE
Is there a way to ONLY look at the source documents?

### DISCUSSION THREAD

**Discussion by langejay** [imartinez/privateGPT#438] [Ideas]
When i ask a simple question involving my source documents, many of my questions take a long time to respond and seems to answer independant of my documents.  Then it will quickly generate an answer regarding my source documents.  It would be good to bypass the (seemingly) first process.

**Comment by Lutios** [comment]
I have the same question. Is there a way to skip the LLM step and just return the sources?

**Reply by mechanicmuthu** [reply]
You need to make some changes in program to do that. You can change in privateGPT.py to call 

`docs = db.as_retriever(search_type="similarity", search_kwargs={"k":4}).get_relevant_documents(query)` 

and totally comment out the part where call is made to retrievalQA 

`qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=retriever, return_source_documents= not args.hide_source)`

It needs some coding though.

**Reply by Lutios** [reply]
It worked, thank you! And now I can control the number of sources returned using search_kwargs={"k":5}. It is also much faster without the LLM inference step.

```python
    ####qa = RetrievalQA.from_chain_type(llm=llm, chain_type="stuff", retriever=retriever, return_source_documents= not args.hide_source)
    # Interactive questions and answers
    while True:
        query = input("\nEnter a query: ")
        if query == "exit":
            break

        # Get the answer from the chain
        ####res = qa(query)
        ####answer, ####
        docs = db.as_retriever(search_type="similarity", search_kwargs={"k":5}).get_relevant_documents(query)

        # # Print the result
        # print("\n\n> Question:")
        # print(query)
        # print("\n> Answer:")
        # print(answer)

        # Print the relevant sources used for the answer
        for document in docs:
            print("\n> " + document.metadata["source"] + ":")
            print(document.page_content)
```

**Reply by chitteshwari** [reply]
@Lutios
How would we get & print the answer in this case since with this improved code, the answer portion is commented out?

**Reply by nicholasqa** [reply]
This seems to only result in the sections of the documents found that are related to the question, but not an actual answer from the LLM. It would be helpful to have an actual answer from the LLM but basing only on the ingested db. Is that possible at all?

For example, I'm ingesting dozens of pdfs but when I ask a specific question that I know to be within those pdfs, I get a generic answer from the LLM that seems to be based on it's "general" knowledge and not on the ingested documents.

**Reply by milincheta** [reply]
this is, as I understand, to point out the relevant documents and parts of them, but not to articulate an answer.

**Comment by NPap0** [comment]
In layman terms,
what you want is to vectorize the database (turn the documents/chunks into numbers/already takes place in privateGPT) then have your question transformed into numbers as well and then you'll use the same similarity function used in privateGPT to fetch the K top similar chunks. Yes it is possible, it shouldn't be too much work // start reading the code and you'll see the part where the top K chunks are returned, you'd like to stop there and return those instead of the answer of the LLM. (I just wanted to rephrase the question so maybe one that has some free time is able to work on it, cheers!)

**Comment by pseudotensor** [comment]
FYI https://github.com/h2oai/h2ogpt supports this now.  Just specify "Only" when selecting subset of documents, and it'll return just the sources it would have used:

![image](https://github.com/imartinez/privateGPT/assets/2249614/79c623bd-a4af-4b4f-b94d-357c281ce0e2)

**Reply by Lutios** [reply]
What's the easiest way to run H20GPT locally on CPU? I was getting a little confused trying to follow your documentation...

**Reply by pseudotensor** [reply]
Setup is similar to privateGPT: https://github.com/h2oai/h2ogpt/blob/main/FAQ.md#cpu

And thanks for feedback about confusing docs.  Will try to improve.

**Reply by pseudotensor** [reply]
The main readme has this: https://github.com/h2oai/h2ogpt/tree/main#cpu and links to the FAQ.md#cpu.  Can you say a bit about what was confusing?  I'm happy to improve anything.  Thanks!

**Reply by pseudotensor** [reply]
Thanks.  I reorganized the CPU installation instructions and put in main readme page.  I think should be less confusing: https://github.com/h2oai/h2ogpt#cpu

**Reply by NPap0** [reply]
> FYI https://github.com/h2oai/h2ogpt supports this now. Just specify "Only" when selecting subset of documents, and it'll return just the sources it would have used:
> 
> ![image](https://user-images.githubusercontent.com/2249614/241065370-79c623bd-a4af-4b4f-b94d-357c281ce0e2.png)

Hey, checked out your work and I'm quite amazed. Got 2 questions if you don't mind:
1. How easy would the implementation be for to go from GUI to python? I would like to use this return specific chunks process as part of a pipeline.
2. What are the differences between PrivateGPT and H2O at this moment? Why should one choose H2O over the other? Just trying to get some pros and cons here. For starters, I think PrivateGPT doesn't get updated as much, maybe the contributors don't have as much time, but I see you are pushing commits on the daily and that is great for further development

**Reply by pseudotensor** [reply]
1. I think you mean a little CLI version like privateGPT instead of GUI or client API. I think that's great point, will do it today.
2. h2oGPT has a UI, supports image captions, one can select which documents out of collection to ask question of, can save any chats, export them, import them, etc.  I should add a list of current differences in our FAQ. Thanks!

**Reply by NPap0** [reply]
> 1. I think you mean a little CLI version like privateGPT instead of GUI or client API. I think that's great point, will do it today.
> 
>     2. h2oGPT has a UI, supports image captions, one can select which documents out of collection to ask question of, can save any chats, export them, import them, etc.  I should add a list of current differences in our FAQ. Thanks!

Hey, just saw your reply. Thank you for continuing your contribution to this! I'll read over the newest commits and definitely give this a try if its available!

**Comment by dgloeppky** [comment]
I have a few more random thoughts on source documents:

A way to delete an ingested file directly from the GUI.

A way to delete multiple document ids from an API endpoint.

A way to quickly clear all embeddings from the database.

A pre-configured watch folder (source_documents) as part of the installation and no need to
run an ingest script.  Any documents placed in the folder would automatically be ingested,
even changes to an already ingested file if it has been modified.
